// @novu/shared/dist/cjs/consts/providers/channels/chat.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chatProviders = void 0;
const credentials_1 = require("../credentials");
const provider_enum_1 = require("../provider.enum");
const types_1 = require("../../../types");
exports.chatProviders = [
    {
        id: provider_enum_1.ChatProviderIdEnum.Slack,
        displayName: 'Slack',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: credentials_1.slackConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/chat/slack',
        logoFileName: { light: 'slack.svg', dark: 'slack.svg' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.Discord,
        displayName: 'Discord',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: [],
        docReference: 'https://docs.novu.co/channels-and-providers/chat/discord',
        logoFileName: { light: 'discord.svg', dark: 'discord.svg' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.GrafanaOnCall,
        displayName: 'Grafana On Call Webhook',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: credentials_1.grafanaOnCallConfig,
        docReference: 'https://grafana.com/docs/oncall/latest/integrations/webhook/',
        logoFileName: { light: 'grafana-on-call.png', dark: 'grafana-on-call.png' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.MsTeams,
        displayName: 'MSTeams',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: [],
        docReference: 'https://docs.novu.co/channels-and-providers/chat/ms-teams',
        logoFileName: { light: 'msteams.svg', dark: 'msteams.svg' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.Mattermost,
        displayName: 'Mattermost',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: [],
        docReference: 'https://developers.mattermost.com/integrate/webhooks/incoming/',
        logoFileName: { light: 'mattermost.svg', dark: 'mattermost.svg' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.Ryver,
        displayName: 'Ryver',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: [],
        docReference: 'https://api.ryver.com/ryvrest_api_examples.html#create-chat-message',
        logoFileName: { light: 'ryver.png', dark: 'ryver.png' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.Zulip,
        displayName: 'Zulip',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: [],
        docReference: 'https://docs.novu.co/channels-and-providers/chat/zulip',
        logoFileName: { light: 'zulip.svg', dark: 'zulip.svg' },
    },
    {
        id: provider_enum_1.ChatProviderIdEnum.GetStream,
        displayName: 'GetStream',
        channel: types_1.ChannelTypeEnum.CHAT,
        credentials: credentials_1.getstreamConfig,
        docReference: 'https://getstream.io/chat/docs/node/?language=javascript',
        logoFileName: { light: 'getstream.svg', dark: 'getstream.svg' },
    },
];
