// @novu/shared/dist/cjs/consts/providers/channels/push.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pushProviders = void 0;
const credentials_1 = require("../credentials");
const provider_enum_1 = require("../provider.enum");
const types_1 = require("../../../types");
exports.pushProviders = [
    {
        id: provider_enum_1.PushProviderIdEnum.OneSignal,
        displayName: 'OneSignal',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.oneSignalConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/onesignal',
        logoFileName: { light: 'one-signal.svg', dark: 'one-signal.svg' },
    },
    {
        id: provider_enum_1.PushProviderIdEnum.Pushpad,
        displayName: 'Pushpad',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.pushpadConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/pushpad',
        logoFileName: { light: 'pushpad.svg', dark: 'pushpad.svg' },
    },
    {
        id: provider_enum_1.PushProviderIdEnum.FCM,
        displayName: 'Firebase Cloud Messaging',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.fcmConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/fcm',
        logoFileName: { light: 'fcm.svg', dark: 'fcm.svg' },
    },
    {
        id: provider_enum_1.PushProviderIdEnum.EXPO,
        displayName: 'Expo Push',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.expoConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/expo-push',
        logoFileName: { light: 'expo.svg', dark: 'expo.svg' },
    },
    {
        id: provider_enum_1.PushProviderIdEnum.APNS,
        displayName: 'APNs',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.apnsConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/apns',
        logoFileName: { light: 'apns.png', dark: 'apns.png' },
        betaVersion: true,
    },
    {
        id: provider_enum_1.PushProviderIdEnum.PushWebhook,
        displayName: 'Push Webhook',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.pushWebhookConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/push-webhook',
        logoFileName: { light: 'push-webhook.svg', dark: 'push-webhook.svg' },
        betaVersion: true,
    },
    {
        id: provider_enum_1.PushProviderIdEnum.PusherBeams,
        displayName: 'Pusher Beams',
        channel: types_1.ChannelTypeEnum.PUSH,
        credentials: credentials_1.pusherBeamsConfig,
        docReference: 'https://docs.novu.co/channels-and-providers/push/pusher-beams',
        logoFileName: { light: 'pusher-beams.svg', dark: 'pusher-beams.svg' },
    },
];
