// @novu/shared/dist/esm/types/feature-flags/flags.types.spec.js
import { testFlagEnumValidity } from './flags.types';
import { FeatureFlagsKeysEnum } from './feature-flags';
import { SystemCriticalFlagsEnum } from './system-critical-flags';
export const validFlag = 'IS_SOMETHING_ENABLED';
export const invalidPrefixFlag = 'SOMETHING_ENABLED';
export const invalidSuffixFlag = 'IS_SOMETHING';
export const invalidSubjectFlag = 'IS_something_ENABLED';
var ValidFlagsEnum;
(function (ValidFlagsEnum) {
    ValidFlagsEnum["IS_SOMETHING_ENABLED"] = "IS_SOMETHING_ENABLED";
    ValidFlagsEnum["IS_SOMETHING_ELSE_ENABLED"] = "IS_SOMETHING_ELSE_ENABLED";
})(ValidFlagsEnum || (ValidFlagsEnum = {}));
testFlagEnumValidity(ValidFlagsEnum);
var InvalidKeyFlagsEnum;
(function (InvalidKeyFlagsEnum) {
    InvalidKeyFlagsEnum["IS_SOMETHING_ENABLED"] = "IS_SOMETHING_ENABLED";
    InvalidKeyFlagsEnum["INVALID_ENABLED"] = "IS_INVALID_ENABLED";
})(InvalidKeyFlagsEnum || (InvalidKeyFlagsEnum = {}));
testFlagEnumValidity(InvalidKeyFlagsEnum);
var InvalidValueFlagsEnum;
(function (InvalidValueFlagsEnum) {
    InvalidValueFlagsEnum["IS_SOMETHING_ENABLED"] = "IS_SOMETHING_ENABLED";
    InvalidValueFlagsEnum["IS_INVALID_ENABLED"] = "INVALID_ENABLED";
})(InvalidValueFlagsEnum || (InvalidValueFlagsEnum = {}));
testFlagEnumValidity(InvalidValueFlagsEnum);
testFlagEnumValidity(FeatureFlagsKeysEnum);
testFlagEnumValidity(SystemCriticalFlagsEnum);
