// @novu/shared/dist/cjs/types/feature-flags/flags.types.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.invalidSubjectFlag = exports.invalidSuffixFlag = exports.invalidPrefixFlag = exports.validFlag = void 0;
const flags_types_1 = require("./flags.types");
const feature_flags_1 = require("./feature-flags");
const system_critical_flags_1 = require("./system-critical-flags");
exports.validFlag = 'IS_SOMETHING_ENABLED';
exports.invalidPrefixFlag = 'SOMETHING_ENABLED';
exports.invalidSuffixFlag = 'IS_SOMETHING';
exports.invalidSubjectFlag = 'IS_something_ENABLED';
var ValidFlagsEnum;
(function (ValidFlagsEnum) {
    ValidFlagsEnum["IS_SOMETHING_ENABLED"] = "IS_SOMETHING_ENABLED";
    ValidFlagsEnum["IS_SOMETHING_ELSE_ENABLED"] = "IS_SOMETHING_ELSE_ENABLED";
})(ValidFlagsEnum || (ValidFlagsEnum = {}));
(0, flags_types_1.testFlagEnumValidity)(ValidFlagsEnum);
var InvalidKeyFlagsEnum;
(function (InvalidKeyFlagsEnum) {
    InvalidKeyFlagsEnum["IS_SOMETHING_ENABLED"] = "IS_SOMETHING_ENABLED";
    InvalidKeyFlagsEnum["INVALID_ENABLED"] = "IS_INVALID_ENABLED";
})(InvalidKeyFlagsEnum || (InvalidKeyFlagsEnum = {}));
(0, flags_types_1.testFlagEnumValidity)(InvalidKeyFlagsEnum);
var InvalidValueFlagsEnum;
(function (InvalidValueFlagsEnum) {
    InvalidValueFlagsEnum["IS_SOMETHING_ENABLED"] = "IS_SOMETHING_ENABLED";
    InvalidValueFlagsEnum["IS_INVALID_ENABLED"] = "INVALID_ENABLED";
})(InvalidValueFlagsEnum || (InvalidValueFlagsEnum = {}));
(0, flags_types_1.testFlagEnumValidity)(InvalidValueFlagsEnum);
(0, flags_types_1.testFlagEnumValidity)(feature_flags_1.FeatureFlagsKeysEnum);
(0, flags_types_1.testFlagEnumValidity)(system_critical_flags_1.SystemCriticalFlagsEnum);
