// @novu/node/build/module/lib/integrations/integrations.js
import { WithHttp } from '../novu.interface';
export class Integrations extends WithHttp {
    async getAll() {
        return await this.http.get('/integrations');
    }
    async getActive() {
        return await this.http.get('/integrations/active');
    }
    async getInAppStatus() {
        return await this.http.get('/integrations/in-app/status');
    }
    async getWebhookProviderStatus(providerId) {
        return await this.http.get(`integrations/webhook/provider/${providerId}/status`);
    }
    async create(providerId, data) {
        return await this.http.post(`/integrations`, {
            providerId,
            ...data,
        });
    }
    async update(integrationId, data) {
        return await this.http.put(`/integrations/${integrationId}`, {
            ...data,
        });
    }
    async setIntegrationAsPrimary(integrationId) {
        return await this.http.post(`/integrations/${integrationId}/set-primary`, {});
    }
    async delete(integrationId) {
        return await this.http.delete(`/integrations/${integrationId}`);
    }
}
